File: backend/app/providers/places/factory.py
"""Selects the nearby-places provider from configuration.

Unlike the other factories this is not cached: the seeded provider is bound to a
request-scoped database session.
"""

from __future__ import annotations

import logging

from sqlalchemy.orm import Session

from app.core.config import settings
from app.providers.places.base import NearbyPlacesProvider
from app.providers.places.overpass import OverpassPlacesProvider
from app.providers.places.seeded import SeededPlacesProvider

logger = logging.getLogger("roadwatch.places")


def get_places_provider(db: Session) -> NearbyPlacesProvider:
    provider = settings.PLACES_PROVIDER.strip().lower()
    if provider == "overpass":
        return OverpassPlacesProvider(db, settings.PLACES_API_URL)
    if provider != "seeded":
        logger.warning("Unknown PLACES_PROVIDER %r; using the seeded catalogue", provider)
    return SeededPlacesProvider(db)

File: supabase/functions/api/providers/places/factory.ts
// Ported from backend/app/providers/places/factory.py. Not cached (like the
// Python version) - the seeded provider is bound to a request-scoped client.

import type { SupabaseClient } from "@supabase/supabase-js";
import { settings } from "../../_shared/config.ts";
import type { NearbyPlacesProvider } from "./base.ts";
import { SeededPlacesProvider } from "./seeded.ts";
import { OverpassPlacesProvider } from "./overpass.ts";

export function getPlacesProvider(client: SupabaseClient): NearbyPlacesProvider {
  const provider = settings.PLACES_PROVIDER.trim().toLowerCase();
  if (provider === "overpass") {
    return new OverpassPlacesProvider(client, settings.PLACES_API_URL);
  }
  return new SeededPlacesProvider(client);
}
